src/providers/wms/qgswmsprovider.cpp

```diff
@@ -888,11 +902,27 @@ QImage *QgsWmsProvider::draw( QgsRectangle const &viewExtent, int pixelWidth, in
 
       if ( QgsTileCache::tile( r.url, localImage ) )
       {
-        double cr = viewExtent.width() / image->width();
-        QRectF dst( ( r.rect.left() - viewExtent.xMinimum() ) / cr,
-                    ( viewExtent.yMaximum() - r.rect.bottom() ) / cr,
-                    r.rect.width() / cr,
-                    r.rect.height() / cr );
+        if ( mConverter && mProviderResamplingEnabled )
+        {
+          if ( sourceResolution < 0 )
+          {
+            sourceResolution = r.rect.width() / localImage.width();
+            int pixelX = std::ceil( viewExtent.width() / sourceResolution ) + 4;
+            int pixelY = std::ceil( viewExtent.height() / sourceResolution ) + 4;
+
+            *image = QImage( pixelX, pixelY, QImage::Format_ARGB32 );
+            image->fill( 0 );
+
+            effectiveViewExtent = QgsRectangle( viewExtent.xMinimum() - 2 * sourceResolution,
+                                                viewExtent.yMinimum() - 2 * sourceResolution,
+                                                viewExtent.xMinimum() + ( image->width() - 2 ) * sourceResolution,
+                                                viewExtent.yMinimum() + ( image->height() - 2 ) * sourceResolution );
```

Comment:
could we replace this just by
```cpp
effectiveViewExtent = viewExtent;
effectiveViewExtent.grow( 2 * sourceResolution );
```
?

Reply:
yes, or better `effectiveViewExtent = viewExtent.buffered( 2 * sourceResolution );`

Reply:
true :)

Reply:
actually no, the initial version return a exact extent considering the pixels counts and the resolution, buffering the initial extent is not exact

Reply:
sorry not sure if I understand...